File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.js
/**
 * @param {string} s
 * @return {string}
 */
var longestPalindrome = function(s) {
  if (s.length === 0) return "";
  if (s.length === 1) return s[0];

  var findPalindrome = function(start, end) {
    while (start >= 0 && end < s.length && s[start] === s[end]) {
      start--;
      end++;
    }
    return s.substring(start + 1, end);
  };

  var evenPalindrome, oddPalindrome;
  for (let i = 0; i < s.length; i++) {
    let even = findPalindrome(i, i + 1);
    let odd = findPalindrome(i, i);
    if (!evenPalindrome || even.length > evenPalindrome.length) {
      evenPalindrome = even;
    }
    if (!oddPalindrome || odd.length > oddPalindrome.length) {
      oddPalindrome = odd;
    }
  }

  return evenPalindrome.length > oddPalindrome.length
    ? evenPalindrome
    : oddPalindrome;
};

File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        n = len(s)
        if n == 0:
            return ""
        
        start, end = 0, 0
        
        dp = [[False] * n for _ in range(n)]
 
        for i in range(n):
            dp[i][i] = True
            start, end = i, i
        
    
        for i in range(n-1, -1, -1):
            for j in range(i+1, n):
                if s[i] == s[j]:
                    if j - i == 1 or dp[i+1][j-1]:
                        dp[i][j] = True
                        if j - i > end - start:
                            start, end = i, j
                            
        return s[start:end+1]
